### DISCUSSION TITLE
Webpack 5 assets modules: raw-loader vs asset/source

### DISCUSSION THREAD

**Discussion by PierBJX** [webpack/webpack#14847] [General]
# Bug report

**What is the current behavior?**
Using `html-webpack-plugin` with a loader for html does not give the same output if I use the loader `raw-loader` or the type `asset/source`.

Using the `raw-loader`, I have a classic _index.html_ corresponding to the template I provide. However, if I use the `asset/source`, I have this kind of output in my _index.html_:

> var _ = non_webpack_require("C:\Users...\node_modules\lodash\lodash.js");module.exports = function (templateParams) { with(templateParams) {return (function(data) { var __t, __p = ''; __p += '<!doctype html>Webpack App'; return __p })();}}

**If the current behavior is a bug, please provide the steps to reproduce.**
To reproduce the issue, just use this _webpack.config.js_ with a _example.js_ and _index.html_ at the root of the project:
```
module.exports = {
  context: __dirname,
  entry: './example.js',
  experiments: {
    asset: true,
  },
  output: {
    path: path.join(__dirname, 'dist'),
    publicPath: '',
    filename: 'bundle.js',
  },
  module: {
    rules: [
      { test: /\.css$/, use: ['style-loader', 'css-loader'] },
      { test: /\.png$/, type: 'asset/resource' },
      { test: /\.html$/, use: ['raw-loader'] },
      // { test: /\.html$/, type: 'asset/source' },
    ],
  },
  plugins: [
    new HtmlWebpackPlugin({
      template: 'index.html',
    }),
  ],
};
```

Switch with the commented code to have the raw-loader or the asset/source and you will notice a different index.html output.

**What is the expected behavior?**

We should have the same output as it is mentioned in the [documentation](https://webpack.js.org/guides/asset-modules).

**Other relevant information:**
webpack version:  5.64.4
Node.js version: 14.17.5
Operating System: Windows 10
Additional tools: 
 - html-webpack-plugin => 5.5.0

**Comment by alexander-akait** [comment]
You should not use `asset/source` for this, why you use it? What is use case?

**Reply by PierBJX** [reply]
In fact, in my entry _example.js_, I import some html files therefore I need a loader to manage them. And I wanted to use the asset/source provided by webpack like in the example in the documentation. However, using it make the output of html-webpack-plugin different and not what I have with raw-loader.
@alexander-akait

**Reply by alexander-akait** [reply]
What is current output with `asset/source`?

**Reply by PierBJX** [reply]
The output with `asset/source` is:

`var _ = __non_webpack_require__("C:\\Users\\PathToProject\\node_modules\\lodash\\lodash.js");module.exports = function (templateParams) { with(templateParams) {return (function(data) { var __t, __p = ''; __p += '<!doctype html>\r\n<html>\r\n<head>\r\n<meta charset="utf-8">\r\n<title>Webpack App</title>\r\n<meta name="viewport" content="width=device-width,initial-scale=1">\r\n<script defer="defer" src="bundle.js"></script></head>\r\n<body></body>\r\n</html>\r\n'; return __p })();}}`

Whereas with `raw-loader`, I have a basic html:
```
<!doctype html>
<html>
<head>
  <meta charset="utf-8">
  <title>Webpack App</title>
  <meta name="viewport" content="width=device-width,initial-scale=1">
  <script defer="defer" src="bundle.js"></script>
</head>
<body></body>
</html>
```

**Reply by alexander-akait** [reply]
hm, I don't see problem... except absolute path to lodash, but it should be bug on html-webpack-plugin (they should context requests), maybe you can provide small repo

**Reply by PierBJX** [reply]
@alexander-akait  You can find this [repo](https://github.com/PierBJX/fluffy-enigma/tree/main/webpack/asset-modules) which reproduces the issue

**Reply by alexander-akait** [reply]
I see... The problem - you want to load html file as `raw` asset, but `html-webpack-plugin` not interpreted it as JS file, so can't execute, bug on `html-webpack-plugin` side (to be honestly it is not bug, rather overlooked possibility), we can't fix it here, please open an issue in `html-webpack-plugin`, ideally they should use `pitch` phase, not normal execution of loaders, so they can add code without problems,

**Reply by alexander-akait** [reply]
As workaround you can use:
```
{ 
  test: /\.html$/,
  // Put here all files which used in `html-webpack-plugin`
  exclude: /index\.html/,
  type: 'asset/source' 
},
```

**Reply by PierBJX** [reply]
Thank you, I opened an [issue](https://github.com/jantimon/html-webpack-plugin/issues/1707) on `html-wepback-plugin`.

**Comment by kara-ryli** [comment]
The root cause of this appears to be that HTMLWebpackPlugin attempts to introspect the loaders being used for the file, and finding none, does its default behavior. The problem is that `asset/source` doesn't show up in the list of loaders, and so HTMLWebpackPlugin doesn't have anything to look for.

Looking at [`LoaderRunnerContext`](https://github.com/webpack/webpack/blob/main/declarations/LoaderContext.d.ts#L87) typedef, I didn't see anything plugin or loader authors could hook into to detect this load method -- adding a `resourceType` property to that object would give the plugin the data it needs to do the right thing.

**Reply by alexander-akait** [reply]
Yes, that sounds resonable, you can send a PR
